src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1277 @@
+
+                _sendBuffer.Discard(_sendBuffer.ActiveLength);
+            }
+        }
+
+        private async ValueTask DrainContentLength0Frames(CancellationToken cancellationToken)
+        {
+            Http3FrameType? frameType;
+            long payloadLength;
+
+            while (true)
+            {
+                (frameType, payloadLength) = await ReadFrameEnvelopeAsync(cancellationToken).ConfigureAwait(false);
+
+                switch (frameType)
+                {
+                    case Http3FrameType.Headers:
+                        // Pick up any trailing headers.
+                        _trailingHeaders = new List<(HeaderDescriptor name, string value)>();
+                        await ReadHeadersAsync(payloadLength, cancellationToken).ConfigureAwait(false);
+
+                        // Stop looping after a trailing header.
+                        // There may be extra frames after this one, but they would all be unknown extension
+                        // frames that can be safely ignored. Just stop reading here.
+                        // Note: this does leave us open to a bad server sending us an out of order DATA frame.
+                        goto case null;
+                    case null:
+                        // Done receiving: shutdown stream and copy over trailing headers.
+                        _stream.ShutdownRead();
+
+                        CopyTrailersToResponseMessage(_response);
+                        return;
+                    case Http3FrameType.Data:
+                        // The sum of data frames must equal content length. Because this method is only
+                        // called for a Content-Length of 0, anything other than 0 here would be an error.
+                        // Per spec, 0-length payload is allowed.
+                        if (payloadLength != 0)
+                        {
+                            if (NetEventSource.IsEnabled)
+                            {
+                                Trace("Response content exceeded Content-Length.");
+                            }
+                            throw new HttpRequestException(SR.net_http_invalid_response);
+                        }
+                        break;
+                    default:
+                        Debug.Fail($"Recieved unexpected frame type {frameType}.");
+                        return;
+                }
+            }
+        }
+
+        private void CopyTrailersToResponseMessage(HttpResponseMessage responseMessage)
+        {
+            if (_trailingHeaders?.Count > 0)
+            {
+                foreach ((HeaderDescriptor name, string value) in _trailingHeaders)
+                {
+                    responseMessage.TrailingHeaders.TryAddWithoutValidation(name, value);
+                }
+                _trailingHeaders.Clear();
+            }
+        }
+
+        private void BufferHeaders(HttpRequestMessage request)
```

Comment:
nit: CommitHeaders? We use this terminology when we write the headers to a buffer and don't flush them.